### PR TITLE
KAFKA-8345: KIP-455: Admin API changes (Part 2)

### PR DESCRIPTION
**Pull request by stanislavkozlovski**
This implements the needed ListPartitionReassignments and  AlterPartitionReassignments AdminClient APIs, KafkaAdminClient implementation and tests for KIP-455

Previous PR: https://github.com/apache/kafka/pull/7114

### CODE REVIEW

File: clients/src/main/java/org/apache/kafka/clients/admin/Admin.java [hunk 1]
```diff
@@ -930,6 +931,120 @@ ElectLeadersResult electLeaders(
         Set<TopicPartition> partitions,
         ElectLeadersOptions options);
 
+
+    /**
+     * Change the reassignments for one or more partitions.
+     * Providing an empty Optional (e.g via {@link Optional#empty()}) will <bold>cancel</bold> the reassignment for the associated partition.
```
Comment (stanislavkozlovski) [inline clients/src/main/java/org/apache/kafka/clients/admin/Admin.java]: maybe we could say it will "reset" the reassignment?

File: clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java [hunk 2]
```diff
@@ -3070,4 +3079,235 @@ void handleFailure(Throwable throwable) {
...
+                                        new TopicPartition(topicName, partition.partitionIndex()),
+                                        new ApiError(topLevelError, topLevelError.message()).exception()
+                                );
+                                receivedResponsesCount += 1;
+                            }
+                        }
+                        break;
+                }
+                if (topLevelError == Errors.NONE)
+                    receivedResponsesCount += validateTopicResponses(response.data().responses(), errors);
```
Comment (viktorsomogyi) [inline clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java]: Why not do this inside the `switch` expr above?

File: clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java [hunk 3]
```diff
@@ -3070,4 +3079,235 @@ void handleFailure(Throwable throwable) {
...
+
+                if (errors.values().stream().noneMatch(Objects::nonNull)
```
Comment (viktorsomogyi) [inline clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java]: nit: This block might be refactored to a method like `assertResponseCountMatch` so `alterPartitionReassignments` becomes shorter.

File: checkstyle/suppressions.xml [hunk 4]
```diff
@@ -23,7 +23,7 @@
               files="AbstractResponse.java"/>
 
     <suppress checks="MethodLength"
-              files="KerberosLogin.java|RequestResponseTest.java|ConnectMetricsRegistry.java|KafkaConsumer.java"/>
+              files="KerberosLogin.java|RequestResponseTest.java|ConnectMetricsRegistry.java|KafkaConsumer.java|KafkaAdminClient.java"/>
```
Comment (viktorsomogyi) [inline checkstyle/suppressions.xml]: I'm not really comfortable with adding exceptions to this check as it often can be avoided.
On a side note I've been thinking about the AdminClient class that it might make sense at some point to slice up the class based on some grouping and delegate these composites in KafkaAdminClient. So we could have like a TopicAdminClient, AclAdminClient, etc. (and of course there might be better ways of grouping, this is just an example).

File: clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java [hunk 5]
```diff
@@ -3070,4 +3079,235 @@ void handleFailure(Throwable throwable) {
...
+                future.completeExceptionally(new InvalidTopicException("The given partition index " +
+                        topicPartition.partition() + " is not valid."));
+            } else {
+                Map<Integer, Optional<NewPartitionReassignment>> partitionReassignments =
+                        topicsToReassignments.get(topicPartition.topic());
+                if (partitionReassignments == null) {
+                    partitionReassignments = new TreeMap<>();
+                    topicsToReassignments.put(topic, partitionReassignments);
+                }
+                partitionReassignments.put(partition, reassignment);
```
Comment (viktorsomogyi) [inline clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java:3128]: This could be more concise:
```
topicsToReassignments.getOrDefault(topicPartition.topic(), new TreeMap<>()).put(partition, reassignment);
```

File: clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java (hunk 2 above)
Comment (stanislavkozlovski) [reply clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java]: No idea why I did that. Thanks

File: clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java (hunk 3 above)
Comment (stanislavkozlovski) [reply clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java]: I can't come up with a way that makes this better to read. Could you give a concrete example?

File: clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java (hunk 5 above)
Comment (stanislavkozlovski) [reply clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java]: I think you meant `putIfAbsent`?

Comment (stanislavkozlovski) [reply clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java:3128]: That wouldn't store the `new TreeMap` in `topicsToReassignments` though, right?

File: checkstyle/suppressions.xml (hunk 4 above)
Comment (stanislavkozlovski) [reply checkstyle/suppressions.xml]: Thanks for the reminder here - we actually don't need this. It was added while developing - the committed changes don't violate the rule

File: clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java (hunk 5 above)
Comment (viktorsomogyi) [reply clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java:3128]: Oh, indeed, I missed that. And in this case you'll need to have an extra line with put.

File: clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java (hunk 3 above)
Comment (viktorsomogyi) [reply clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java]: It would be something like this:
```
private void assertResponseCountMatch(Map<TopicPartition, ApiException> errors, int expectedResponsesCount, int receivedResponsesCount) {
                if (errors.values().stream().noneMatch(Objects::nonNull)
                        && receivedResponsesCount != expectedResponsesCount) {
                    String quantifier = receivedResponsesCount > expectedResponsesCount ? "many" : "less";
                    throw new UnknownServerException("The server returned too " + quantifier + " results." +
                            "Expected " + expectedResponsesCount + " but received " + receivedResponsesCount);
                }
            }
```

File: clients/src/main/java/org/apache/kafka/clients/admin/Admin.java [hunk 6]
```diff
@@ -930,6 +931,120 @@ ElectLeadersResult electLeaders(
...
+     * <p>The following exceptions can be anticipated when calling {@code get()} on the futures obtained from
+     * the returned {@code AlterPartitionReassignmentsResult}:</p>
+     * <ul>
+     *   <li>{@link org.apache.kafka.common.errors.ClusterAuthorizationException}
+     *   If the authenticated user didn't have alter access to the cluster.</li>
+     *   <li>{@link org.apache.kafka.common.errors.UnknownTopicOrPartitionException}
+     *   If the topic or partition does not exist within the cluster.</li>
+     *   <li>{@link org.apache.kafka.common.errors.TimeoutException}
+     *   if the request timed out before the controller could record the new assignments.</li>
+     *   <li>{@link org.apache.kafka.common.errors.InvalidReplicaAssignmentException}
```
Comment (stanislavkozlovski) [inline clients/src/main/java/org/apache/kafka/clients/admin/Admin.java:960]: I think we should opt out of this excepiton. In my controller changes, I haven't implemented it -- mainly because it's tricky to know when a replica is intermittently offline, as the znode and controller memory cache store only the online brokers.

File: clients/src/main/java/org/apache/kafka/clients/admin/Admin.java (hunk 1 above)
Comment (cmccabe) [reply clients/src/main/java/org/apache/kafka/clients/admin/Admin.java]: hmm, cancel seems like a better description?

File: clients/src/main/java/org/apache/kafka/clients/admin/Admin.java (hunk 6 above)
Comment (cmccabe) [reply clients/src/main/java/org/apache/kafka/clients/admin/Admin.java:960]: Can you think of a scenario where we'd want to create a reassignment to a broker that isn't up?  It seems like most (all?) of the time, this would be a mistake.

Another reason why the assignment could be invalid is if the same broker ID appeared more than once (reassignment to 1, 1, 1, etc.)

**Comment by cmccabe** [comment]
LGTM overall.  I left a few comments

File: clients/src/main/java/org/apache/kafka/clients/admin/Admin.java (hunk 1 above)
Comment (stanislavkozlovski) [reply clients/src/main/java/org/apache/kafka/clients/admin/Admin.java]: Right, `reset` isn't a good word - I was thinking of `revert` when I typed it. I don't have a strong opinion on this

File: clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java (hunk 3 above)
Comment (stanislavkozlovski) [reply clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java]: Ah, just that block? Sure, let's do it.

I generally like encapsulating small chunks like this into methods because it makes the code waay easier to read.
Yet, I believe I've had it questioned in some reviews in terms of "why have this in a method if it's used only once"

Glad to have the opposite recommended :)

File: clients/src/main/java/org/apache/kafka/clients/admin/Admin.java (hunk 6 above)
Comment (stanislavkozlovski) [reply clients/src/main/java/org/apache/kafka/clients/admin/Admin.java:960]: Come to think of it... I can't. I think it makes sense to have as long as we do the validation on API calls directly

**Comment by stanislavkozlovski** [comment]
Test failures seem unrelated:
JDK 11/Scala 2.12
`org.apache.kafka.streams.integration.QueryableStateIntegrationTest.queryOnRebalance` - streams test

JDK 8/Scala 2.11
`org.apache.kafka.connect.integration.ExampleConnectIntegrationTest.testSourceConnector` - connect test
`kafka.admin.ResetConsumerGroupOffsetTest.testResetOffsetsExistingTopic`
```
java.util.concurrent.ExecutionException: org.apache.kafka.common.errors.CoordinatorNotAvailableException: The coordinator is not available.
```

File: clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java (hunk 3 above)
Comment (viktorsomogyi) [reply clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java]: Yea I can understand the argument as I often debate it during development but the JIT compiler would inline these kind of methods so you actually don't lose any performance yet the code looks more understandable :)

Comment (stanislavkozlovski) [reply clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java]: I don't think performance was the reason I've got such remarks but yeah, I agree it looks better

File: clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java [hunk 7]
```diff
@@ -1390,9 +1403,7 @@ void handleResponse(AbstractResponse abstractResponse) {
                 // Check for controller change
                 for (Errors error : response.errorCounts().keySet()) {
                     if (error == Errors.NOT_CONTROLLER) {
-                        metadataManager.clearController();
-                        metadataManager.requestUpdate();
-                        throw error.exception();
+                        handleNotControllerError(error);
```
Comment (cmccabe) [inline clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java]: This seems like an unrelated change that should be in a separate PR?

File: clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java [hunk 8]
```diff
@@ -1315,9 +1330,7 @@ public void handleResponse(AbstractResponse abstractResponse) {
...
```
Comment (cmccabe) [inline clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java]: This seems like an unrelated change that should be in a separate PR?

File: clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java [hunk 9]
```diff
@@ -2252,9 +2263,7 @@ public void handleResponse(AbstractResponse abstractResponse) {
                 // Check for controller change
                 for (ApiError error : response.errors().values()) {
                     if (error.error() == Errors.NOT_CONTROLLER) {
-                        metadataManager.clearController();
-                        metadataManager.requestUpdate();
-                        throw error.exception();
+                        handleNotControllerError(error.error());
```
Comment (cmccabe) [inline clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java]: Not related

File: clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java [hunk 10]
```diff
@@ -3070,4 +3079,237 @@ void handleFailure(Throwable throwable) {
...
+            for (TopicPartition tp : partitions.get()) {
+                String topic = tp.topic();
+                int partition = tp.partition();
+                if (topicNameIsUnrepresentable(topic)) {
+                    partitionReassignmentsFuture.completeExceptionally(new InvalidTopicException("The given topic name '"
+                            + topic + "' cannot be represented in a request."));
+                } else if (partition < 0) {
+                    partitionReassignmentsFuture.completeExceptionally(new InvalidTopicException("The given partition index " +
+                            partition + " is not valid."));
+                }
```
Comment (cmccabe) [inline clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java:3262]: The unrepresentable or invalid topics still get sent later, right?  We need a way of making sure that they don't (perhaps by checking whether the future has already been completed?)

File: clients/src/main/java/org/apache/kafka/clients/admin/PartitionReassignment.java [hunk 11]
```diff
@@ -0,0 +1,59 @@
...
+/**
+ * A partition reassignment, which has been listed via {@link AdminClient#listPartitionReassignments()}.
+ */
+public class PartitionReassignment {
+
+    private final List<Integer> replicas;
+    private final List<Integer> addingReplicas;
+    private final List<Integer> removingReplicas;
+
+    public PartitionReassignment(List<Integer> replicas, List<Integer> addingReplicas, List<Integer> removingReplicas) {
```
Comment (cmccabe) [inline clients/src/main/java/org/apache/kafka/clients/admin/PartitionReassignment.java:32]: It would be nice to use Collections.unmodifiableList here for these mutable lists.

File: clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java (hunk 9 above)
Comment (stanislavkozlovski) [reply clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java]: Oops, this is a bug - it should be `handleNotControllerError(error)`

File: clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java (hunk 10 above)
Comment (stanislavkozlovski) [reply clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java:3262]: Oops, nice catch

**Comment by stanislavkozlovski** [comment]
re: `handleNotControllerError` I saw an opportunity to refactor and thought I'd reduce duplication throughout. I'll just use this method in the two new methods introduced by this PR, then

File: clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java [hunk 12]
```diff
@@ -3082,6 +3097,240 @@ void handleFailure(Throwable throwable) {
...
+            }
+
+            @Override
+            void handleFailure(Throwable throwable) {
+                for (KafkaFutureImpl<Void> future : futures.values()) {
+                    future.completeExceptionally(throwable);
+                }
+            }
+        };
+        if (!reassignments.isEmpty()) {
```
Comment (cmccabe) [inline clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java]: Even if reassignments isn't empty, topicsToReassignments may be empty, if all the inputs are invalid.  We should be checking topicsToReassignments here, not reassignments.

Comment (stanislavkozlovski) [reply clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java]: again, great catch. thanks so much

**Comment by cmccabe** [comment]
LGTM
